WalletWasabi.Tests/RegTests.cs

```diff
@@ -1135,6 +1136,133 @@ private void WalletTestsAsync_MemPoolService_TransactionReceived(object sender,
+
+			// Generate some coins
+			await Global.RpcClient.GenerateAsync(2);
+
+			try
+			{
+				nodes.Connect(); // Start connection service.
+				node.VersionHandshake(); // Start mempool service.
+				indexDownloader.Synchronize(requestInterval: TimeSpan.FromSeconds(3)); // Start index downloader service.
+
+				// Wait until the filter our previous transaction is present.
+				var blockCount = await Global.RpcClient.GetBlockCountAsync();
+				await WaitForFiltersToBeProcessedAsync(TimeSpan.FromSeconds(120), blockCount);
+
+				using (var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30)))
+				{
+					await wallet.InitializeAsync(cts.Token); // Initialize wallet service.
+				}
+				Assert.Single(wallet.Coins);
+
+				// Send money before reorg.
+				var operations = new[]{
+					new WalletService.Operation(scp, Money.Coins(0.5m), "") };
+				var btx = await wallet.BuildTransactionAsync("password", operations, 2);
+
+				operations = new[]{
+					new WalletService.Operation(scp, Money.Coins(0.5m), "") };
+				btx = await wallet.BuildTransactionAsync("password", operations, 2);
+
+				// Test synchronization after fork.
+				// Invalidate the blocks containing the funding transaction
+				var tip = await Global.RpcClient.GetBestBlockHashAsync();
+				await Global.RpcClient.InvalidateBlockAsync(tip); // Reorg 1
+				tip = await Global.RpcClient.GetBestBlockHashAsync();
+				await Global.RpcClient.InvalidateBlockAsync(tip); // Reorg 2
+
+				// Generate two new blocks (replace the previous invalidated ones)
+				_filtersProcessedByWalletCount = 0;
+				await Global.RpcClient.GenerateAsync(3);
+				await WaitForFiltersToBeProcessedAsync(TimeSpan.FromSeconds(120), 3);
+
+				// Send money after reorg.
+				// When we invalidate a block, those transactions setted in the invalidated block
+				// are reintroduced when we generate a new block though the rpc call 
+				operations = new[]{
+					new WalletService.Operation(scp, Money.Coins(0.5m), "") };
+				btx = await wallet.BuildTransactionAsync("password", operations, 2);
+
+				operations = new[]{
+					new WalletService.Operation(scp, Money.Coins(0.5m), "") };
+				btx = await wallet.BuildTransactionAsync("password", operations, 2);
+
+				
+				// Test synchronization after fork with different transactions.
+				// Create a fork that invalidates the blocks containing the funding transaction
+				_filtersProcessedByWalletCount = 0;
+				var winningFork = await RegTestFixture.BackendRegTestNode.GenerateEmptyBlocksAsync(100,
+					new Key().PubKey.GetAddress(Global.RpcClient.Network), 10);
+
+				tip = await Global.RpcClient.GetBestBlockHashAsync();
+				Assert.Equal(tip, winningFork.Last().GetHash());
+				
+				await WaitForFiltersToBeProcessedAsync(TimeSpan.FromSeconds(120), 10);
+
+				// When we creates a winning fork, those transactions setted in the orphand blocks
+				// do not exist any more
+				var coin = wallet.Coins.First();
```

Comment:
I believe if you reorg, the `Coins` list will purge the transaction, not make it unconfirmed. We only make a transaction unconfirmed when a node propagates it to us.

Btw, the problem with light wallets is someone can send fake unconfirmed transactions, even if the inputs are spent previously. Since we don't keep an utxo set we cannot make sure that transaction is not double spent, so we just accept it.